pkg/auth/oauth2_client.go

```diff
@@ -0,0 +1,191 @@
+)
+
+// The following provides the setup required for the client to perform the "Authorization Code" flow with PKCE in order
+// to obtain an access token for public/untrusted clients.
+
+const cookiePKCE = "isPKCE"
+
+var (
+	// pkceCodeVerifier stores the generated random value which the client will on-send to the auth server with the received
+	// authorization code. This way the oauth server can verify that the base64URLEncoded(sha265(codeVerifier)) matches
+	// the stored code challenge, which was initially sent through with the code+PKCE authorization request to ensure
+	// that this is the original user-agent who requested the access token.
+	PkceCodeVerifier string
+
+	// pkceCodeChallenge stores the base64(sha256(codeVerifier)) which is sent from the
+	// client to the auth server as required for PKCE.
+	PkceCodeChallenge string
+
+
+	TokenChannel chan string
+)
+
+// The following sets up the requirements for generating a standards compliant PKCE code verifier.
+const codeVerifierLenMin = 43
+const codeVerifierLenMax = 128
+const codeVerifierAllowedLetters = "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ._~"
+
+// generateCodeVerifier provides an easy way to generate an n-length randomised
+// code verifier.
+func generateCodeVerifier(n int) string {
+	// Enforce standards compliance...
+	if n < codeVerifierLenMin {
+		n = codeVerifierLenMin
+	}
+	if n > codeVerifierLenMax {
+		n = codeVerifierLenMax
+	}
+
+	// Randomly choose some allowed characters...
+	b := make([]byte, n)
+	for i := range b {
+		// ensure we use non-deterministic random ints.
+		j, _ := rand.Int(rand.Reader, big.NewInt(int64(len(codeVerifierAllowedLetters))))
+		b[i] = codeVerifierAllowedLetters[j.Int64()]
+	}
+
+	return string(b)
+}
+
+// generateCodeChallenge returns a standards compliant PKCE S(HA)256 code
+// challenge.
+func generateCodeChallenge(codeVerifier string) string {
+	// Create a sha-265 hash from the code verifier...
+	s256 := sha256.New()
+	s256.Write([]byte(codeVerifier))
+
+	// Then base64 encode the hash sum to create a code challenge...
+	return base64.RawURLEncoding.EncodeToString(s256.Sum(nil))
+}
+
+// resetPKCE cleans up PKCE details and returns the code verifier.
+func resetPKCE() (codeVerifier string) {
+	codeVerifier = PkceCodeVerifier
+	PkceCodeVerifier = ""
+	return codeVerifier
+}
+
+
+// A valid oauth2 client (check the store) that additionally requests an OpenID Connect id token
+var clientConf = goauth.Config{
+	ClientID:     "flytectl",
+	ClientSecret: "foobar",
+	RedirectURL:  "http://localhost:3846/callback",
+	Scopes:       []string{"all"},
+	Endpoint: goauth.Endpoint{
+		TokenURL: "http://localhost:8088/oauth2/token",
+		AuthURL:  "http://localhost:8088/oauth2/authorize",
+	},
+}
+
+type FlyteCtlTokenSource struct {
+	accessToken string
+}
+
+func (ts *FlyteCtlTokenSource) Token() (*goauth.Token, error) {
+	t := &goauth.Token{
+		AccessToken: ts.accessToken,
+		Expiry:      time.Now().Add(1 * time.Minute),
```

Comment:
We should read that from the access token response... admin sends the expiry, right? if not I can fix that..

Reply:
Yeah.  i have fixed this by using the tokenChannel to pass in the entire object instead of just the accessToken.
In that way we have all the details of the token and i don't have to create a seperate copy of it.

Have created two channels,
1] error channel  used for communicating errors after the callback handler is called.
2] token channel used for communicating the token